examples/hmm.py

```diff
@@ -0,0 +1,286 @@
+#
+# This model includes two iaranges: one for minibatches of data, and one
+# for the data_dim = 88 keys on the piano. This model has two "style" parameters
+# probs_x and probs_y that we'll draw from a prior. The latent state is x,
+# and the observed state is y. We'll drive probs_* with the guide, enumerate
+# over x, and condition on y.
+#
+# Importantly, the dependency structure of the enumerated variables has
+# narrow treewidth, therefore admitting efficient inference by message passing.
+# Pyro's TraceEnum_ELBO will find an efficient message passing scheme if one
+# exists.
+def model_1(sequences, lengths, args, batch_size=None, include_prior=True):
+    num_sequences, max_length, data_dim = sequences.shape
+    assert lengths.shape == (num_sequences,)
+    assert lengths.max() <= max_length
+    with poutine.mask(mask=torch.tensor(include_prior)):
+        probs_x = pyro.sample("probs_x",
+                              dist.Dirichlet(0.9 * torch.eye(args.hidden_dim) + 0.1)
+                                  .independent(1))
+        probs_y = pyro.sample("probs_y",
+                              dist.Beta(0.1, 0.9)
+                                  .expand([args.hidden_dim, data_dim])
+                                  .independent(2))
+    tones_iarange = pyro.iarange("tones", data_dim, dim=-1)
+    with pyro.iarange("sequences", len(sequences), batch_size, dim=-2) as batch:
+        lengths = lengths[batch]
+        x = 0
+        for t in range(lengths.max()):
+            with poutine.mask(mask=(t < lengths).unsqueeze(-1)):
+                x = pyro.sample("x_{}".format(t), dist.Categorical(probs_x[x]),
+                                infer={"enumerate": "parallel", "expand": False})
+                with tones_iarange:
+                    pyro.sample("y_{}".format(t), dist.Bernoulli(probs_y[x]),
+                                obs=sequences[batch, t])
+
+
+# Next let's add a dependency of y[t] on y[t-1].
+#
+#     x[t-1] --> x[t] --> x[t+1]
+#        |        |         |
+#        V        V         V
+#     y[t-1] --> y[t] --> y[t+1]
+#
+def model_2(sequences, lengths, args, batch_size=None, include_prior=True):
+    num_sequences, max_length, data_dim = sequences.shape
+    assert lengths.shape == (num_sequences,)
+    assert lengths.max() <= max_length
+    with poutine.mask(mask=torch.tensor(include_prior)):
+        probs_x = pyro.sample("probs_x",
+                              dist.Dirichlet(0.9 * torch.eye(args.hidden_dim) + 0.1)
+                                  .independent(1))
+        probs_y = pyro.sample("probs_y",
+                              dist.Beta(0.1, 0.9)
+                                  .expand([args.hidden_dim, 2, data_dim])
+                                  .independent(3))
+    tones_iarange = pyro.iarange("tones", data_dim, dim=-1)
+    with pyro.iarange("sequences", len(sequences), batch_size, dim=-2) as batch:
+        lengths = lengths[batch]
+        x, y = 0, 0
```

Comment:
is it useful to do something like in the dmm? `x_prev =0 .... x_prev = x`? more lines of code but clarifies the structure. `x=pyro.sample(...x...)` might be confusing to the pyro beginner